packages/dx-react-grid/src/components/table-layout/virtual-row-layout.tsx

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import { argumentsShallowEqual } from '@devexpress/dx-core';
+import { VirtualRowLayoutProps } from '../../types';
+
+const reducer = (acc, { column, colSpan }) => {
+  const [columns, colSpans] = acc;
+  columns.push(column);
+  colSpans.push(colSpan);
+  return acc;
+};
+
+export class VirtualRowLayout extends React.Component<VirtualRowLayoutProps> {
+  shouldComponentUpdate(nextProps) {
+    const { cells: prevCells, row: prevRow } = this.props;
+    const { cells: nextCells, row: nextRow } = nextProps;
+    const [nextColumns, nextColSpans] = nextCells.reduce(reducer, [[], []]);
+    const [prevColumns, prevColSpans] = prevCells.reduce(reducer, [[], []]);
+
+    const propsAreEqual = argumentsShallowEqual(nextColumns, prevColumns)
+      && argumentsShallowEqual(nextColSpans, prevColSpans)
+      && prevRow === nextRow;
```

Comment:
Since it is all about performance the `prevRow === nextRow` check should be moved to the beginning of the method.

```js
if (prevRow !== nextRow) {
  return true;
}
```

It would allow to avoid unnecessary arrays creation and comparing.